CPDShell/Core/algorithms/ClassificationBasedCPD/classifiers/knn/knn_classifier.py

```diff
@@ -0,0 +1,131 @@
+"""
+Module for implementation of classifier based on nearest neighbours for cpd.
+"""
+
+__author__ = "Artemii Patov"
+__copyright__ = "Copyright (c) 2024 Artemii Patov"
+__license__ = "SPDX-License-Identifier: MIT"
+
+import typing as tp
+from collections.abc import Iterable
+from math import sqrt
+
+import numpy as np
+
+from CPDShell.Core.algorithms.ClassificationBasedCPD.abstracts.iclassifier import Classifier
+from CPDShell.Core.algorithms.ClassificationBasedCPD.classifiers.knn.knn_graph import KNNGraph
+
+
+class KNNAlgorithm(Classifier):
+    """
+    The class implementing classifier based on nearest neighbours.
+    """
+
+    def __init__(
+        self,
+        metric: tp.Callable[[float, float], float] | tp.Callable[[np.float64, np.float64], float],
+        k=3,
+        delta: float = 1e-12,
+    ) -> None:
+        """
+        Initializes a new instance of KNN classifier for cpd.
+
+        :param metric: function for calculating distance between points in time series.
+        :param k: number of neighbours in graph relative to each point.
+        :param delta: delta for comparing floats.
+        """
+        self.__k = k
+        self.__metric = metric
+        self.__delta = delta
+
+        self.__window: list[float | np.float64] = None
+        self.__knn_graph: KNNGraph | None = None
+
+    def classify(self, window: Iterable[float | np.float64]) -> None:
+        """Applies classificator to the given sample.
+
+        :param window: part of global data for finding change points.
+        """
+        self.__window = list(window)
+        self.__knn_graph = KNNGraph(window, self.__metric, self.__k, self.__delta)
```

Comment:
So your algorithm takes a dataset every time and builds a graph using it inside the algorithm? Why is it better not to immediately transfer the graph to the algorithm class? :fishsticks: